### DISCUSSION TITLE
Plans to make "low level" rustic API available for developers?

### DISCUSSION THREAD

**Discussion by czeidler** [rustic-rs/rustic#1171] [Q&A] [answered]
I started working on an idea for a shared data storage solution which uses restic/rustic to do all the heavy lifting of syncing, encrypting and managing data snapshots. However, it would be very useful to me if I could use some of the private rustic structs like the `Packer` or `FileArchiver` to do some direct `Tree` manipulations. I wonder if in general it would be possible to make some of the "low level" functionality available? or is this not a goal of rustic?

For example, one of my ideas is to mount a snapshot as a file system with a writeable overlay layer. When a user updates/adds/deletes a file in the mounted snapshot the file changes are saved locally. The user can then PUSH these local changes to the repository as a new snapshot. This could, for example, be useful to add files to a large snapshot that doesn't fit onto your local device. Without access to the low level rustic APIs I would have to re-implement a fair bit of logic.

**Comment by aawsome** [comment, marked as answer]
Thanks for opening this discussion @czeidler!

In fact there are (at least) two reasons why we didn't make low level API public in rustic_core:
- The project is in an early phase and subject to refactorings. Those for sure will change low level APIs.
- We want to prevent users of the library from doing things wrong.

Saying this, I can think of publishing lower-level APIs under a feature flag - under the explicit warning that those things will change and may even change dramatically in future.

I am very short on time currently, but if you are willing to propose a PR I definitively will support that.

**Reply by czeidler** [reply]
thanks this makes great sense to me. I am also short on time so don't expect anything soon. However, with this knowledge I will keep going with a local copy of rustic and see which APIs I need and how they can be exported.
